src/System.Private.CoreLib/shared/System/Number.Formatting.cs

```diff
@@ -378,92 +378,186 @@ public static bool TryFormatDouble(double value, ReadOnlySpan<char> format, Numb
+                    // also always formats a single integral digit, so we need to increase the precision
+                    // specifier and treat it as the number of significant digits to account for this.
+
+                    if (precision == -1)
                     {
-                        DoubleToNumber(value, 17, ref number);
-                        NumberToString(ref sb, ref number, 'G', 17, info);
+                        precision = 6;
                     }
 
-                    return null;
+                    precision++;
+                    isSignificantDigits = true;
+
+                    break;
                 }
 
-                case 'E':
-                case 'e':
-                    // Round values less than E14 to 15 digits
-                    if (digits > 14)
+                case 'F':
+                case 'f':
+                case 'N':
+                case 'n':
+                {
+                    // The fixed-point and number formats use the precision specifier to indicate the number 
+                    // of decimal digits to format. This defaults to NumberFormatInfo.NumberDecimalDigits.
+
+                    if (precision == -1)
                     {
-                        precision = 17;
+                        precision = info.NumberDecimalDigits;
                     }
+                    isSignificantDigits = false;
+
                     break;
+                }
 
                 case 'G':
                 case 'g':
-                    // Round values less than G15 to 15 digits. G16 and G17 will not be touched.
-                    if (digits > 15)
+                {
+                    // The general format uses the precision specifier to indicate the number of significant
+                    // digits to format. This defaults to the shortest roundtrippable string. Additionally,
+                    // given that we can't return zero significant digits, we treat 0 as returning the shortest
+                    // roundtrippable string as well.
+
+                    if (precision == 0)
                     {
-                        precision = 17;
+                        precision = -1;
                     }
+                    isSignificantDigits = true;
+
                     break;
+                }
+
+                case 'P':
+                case 'p':
+                {
+                    // The percent format uses the precision specifier to indicate the number of
+                    // decimal digits to format. This defaults to NumberFormatInfo.PercentDecimalDigits.
+                    // However, the percent format also always multiplies the number by 100, so we need
+                    // to increase the precision specifier to ensure we get the appropriate number of digits.
+
+                    if (precision == -1)
+                    {
+                        precision = info.PercentDecimalDigits;
+                    }
+
+                    precision += 2;
```

Comment:
Why do we increase it by 2 here? Multiplying by 100?

Reply:
Yes. The percent specifier uses precision to specify the number of decimal digits that are displayed. It also "multiplies" the value by 100 before displaying it, so we need up to 2 additional digits to ensure the user always gets the requested number of decimal digits.